### PR TITLE
Added Confusion to be healed with Full Heals and Full Restores

### PR DESCRIPTION
**Pull request by EvasiveAce**
# Fixes Issue #1054 

Please let me know if there is anything I could improve.

Video testing for Full Heal:

[Video](https://github.com/pagefaultgames/pokerogue/assets/71776311/5f94ef43-c060-4ba4-8fd4-f17b0f9876d9)

Video testing for Full Restore:

[Video](https://github.com/pagefaultgames/pokerogue/assets/71776311/21fda8de-51af-4569-b8db-9290e9064569)

Video testing that it still will have no effect if nothing is wrong:

[Video](https://github.com/pagefaultgames/pokerogue/assets/71776311/2fd8d315-a23a-43e0-95e8-86f04b43c941)

### CODE REVIEW

File: src/field/pokemon.ts [hunk 1]
```diff
@@ -2160,6 +2160,8 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
       if (this.getTag(BattlerTagType.NIGHTMARE))
         this.lapseTag(BattlerTagType.NIGHTMARE);
     }
+    if (this.getTag(BattlerTagType.CONFUSED))
+      this.lapseTag(BattlerTagType.CONFUSED)
```
Comment (CodeTappert) [inline src/field/pokemon.ts:2164]: Please add ; here at the end of 2164

Comment (EvasiveAce) [reply src/field/pokemon.ts]: Done.

**Comment by torranx** [comment]
Are there any scenarios where: 
- pokemon is confused
- resetStatus() is called
- what called the resetStatus() is _not_ Full Heals nor Full Restores

**Comment by rayanzmn** [comment]
@torranx is right, for example the move heal bell which is supposed to only heal non-volatile status conditions (sleep, poison, paralysis, freeze and burn) would also heal confusion due to how resetStatus() was changed.

**Comment by EvasiveAce** [comment]
@torranx and @rayanzmn you both are completely right. Thank you for bringing it to my attention.

I have some video testing that should show that it is fixed.

Video to show that Heal Bell does not affect confusion:

[Video](https://github.com/pagefaultgames/pokerogue/assets/71776311/bba066ae-e0f5-4b46-b662-5799a3d60e06)

Video to show that Full Heals include confusion, as well as not allowing further use once confusion is fixed:

[Video](https://github.com/pagefaultgames/pokerogue/assets/71776311/8bb9a3f6-04d1-48e0-8474-80f2c862cd00)

Video to show that Full Restores include confusion at max HP:

[Video](https://github.com/pagefaultgames/pokerogue/assets/71776311/3180155e-f349-46f5-82ab-0e8395a68f57)

Video to show that Full Restores include confusion at any HP:

[Video](https://github.com/pagefaultgames/pokerogue/assets/71776311/6a66d42b-d2df-44f6-9b7c-af4d5e68b52c)

File: src/field/pokemon.ts [hunk 2]
```diff
@@ -2160,6 +2161,10 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
       if (this.getTag(BattlerTagType.NIGHTMARE))
         this.lapseTag(BattlerTagType.NIGHTMARE);
     }
+    if(confusion) {
```
Comment (torranx) [inline src/field/pokemon.ts:2164]: nit: space after if

File: src/modifier/modifier.ts [hunk 3]
```diff
@@ -1035,9 +1035,8 @@ export class PokemonHpRestoreModifier extends ConsumablePokemonModifier {
       if (!this.fainted)
         restorePoints = Math.floor(restorePoints * (args[1] as number));
       if (this.fainted || this.healStatus)
-        pokemon.resetStatus();
+        pokemon.resetStatus(true, true)
```
Comment (torranx) [inline src/modifier/modifier.ts:1038]: nit: missing semicolon

**Comment by ArchSinccubus** [comment]
Tested new resetStatus() method with both Confusion and a non volatile status (burn in this case)
chrome_wVp81VLQFB

**Comment by stendler** [comment]
Ah yes, the case with Heal Bell was what I suspected in the issue, but not really sure about.

@EvasiveAce Lum Berry already has some code to also heal Confusion. I think, this could now be refactored to  use the new parameter on resetStatus?

**Comment by EvasiveAce** [comment]
For the Lum Berry refactor, it seemed like this was the easiest way to do it.
@stendler if you want to see the refactor.

Video that shows the Lum Berry still heals confusion:

[Video](https://github.com/pagefaultgames/pokerogue/assets/71776311/a3b2ab17-0aba-49b2-a724-915228f84fc2)

Video that shows the Lum Berry still heals non-volatile status conditions:
 
[Video](https://github.com/pagefaultgames/pokerogue/assets/71776311/7989db65-dea4-45a6-87dc-06d04f3555d6)

File: src/field/pokemon.ts [hunk 4]
```diff
@@ -2395,6 +2396,11 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
         this.lapseTag(BattlerTagType.NIGHTMARE);
       }
     }
+    if (confusion) {
+      if (this.getTag(BattlerTagType.CONFUSED)) {
+        this.lapseTag(BattlerTagType.CONFUSED);
```
Comment (Tempo-anon) [inline src/field/pokemon.ts:2401]: Doesn't lapseTag only make it go through a single turn? Shouldn't this be removeTag?

Comment (EvasiveAce) [reply src/field/pokemon.ts:2401]: Either works, not sure why there  are two in general.

**Review by Tempo-anon — APPROVED** [review]

**Pull request merged**